File: datalib/src/main/java/nl/acidcats/tumblrlikes/data_impl/photodata/gateway/greendao/filters/AbstractFilterOptionImpl.java
package nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.filters;

import android.support.annotation.Nullable;

import org.greenrobot.greendao.query.CountQuery;
import org.greenrobot.greendao.query.QueryBuilder;

import java.util.List;

import nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.entities.PhotoEntity;
import nl.acidcats.tumblrlikes.db_impl_greendao.PhotoEntityDao;

/**
 * Created by stephan on 17/05/2017.
 */

abstract class AbstractFilterOptionImpl implements FilterOption {
    private static final String TAG = AbstractFilterOptionImpl.class.getSimpleName();

    PhotoEntityDao _photoEntityDao;

    private CountQuery<PhotoEntity> _countQuery;

    void setPhotoEntityDao(PhotoEntityDao photoEntityDao) {
        _photoEntityDao = photoEntityDao;

        _countQuery = getQueryBuilder().buildCount();
    }

    @Override
    public int getCount() {
        return (int)_countQuery.count();
    }

    @Nullable
    @Override
    public PhotoEntity getPhoto(int index) {
        return getQueryBuilder().limit(1).offset(index).unique();
    }

    protected abstract QueryBuilder<PhotoEntity> getQueryBuilder();
}

File: datalib/src/main/java/nl/acidcats/tumblrlikes/data_impl/photodata/gateway/greendao/filters/AbstractFilterOptionImpl.kt
package nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.filters

import nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.entities.PhotoEntity
import org.greenrobot.greendao.query.CountQuery
import org.greenrobot.greendao.query.QueryBuilder

/**
 * Created on 25/10/2018.
 */
abstract class AbstractFilterOptionImpl : FilterOption {

    override val count
        get() = queryBuilder.count().toInt()

    override fun getPhoto(index: Int): PhotoEntity = queryBuilder.limit(1).offset(index).build().forCurrentThread().unique()

    protected abstract val queryBuilder: QueryBuilder<PhotoEntity>
}
